pkg/ccl/changefeedccl/sink_kafka_v2.go

```diff
@@ -0,0 +1,630 @@
+		opts = append(opts, kgo.MaxBufferedRecords(sinkCfg.Flush.MaxMessages))
+	}
+
+	if sinkCfg.ClientID != "" {
+		opts = append(opts, kgo.ClientID(sinkCfg.ClientID))
+	}
+
+	switch strings.ToUpper(sinkCfg.RequiredAcks) {
+	case ``, `ONE`, `1`: // This is our default.
+		opts = append(opts, kgo.RequiredAcks(kgo.LeaderAck()))
+	case `ALL`, `-1`:
+		opts = append(opts, kgo.RequiredAcks(kgo.AllISRAcks()))
+	case `NONE`, `0`:
+		opts = append(opts, kgo.RequiredAcks(kgo.NoAck()))
+	default:
+		return nil, errors.Errorf(`unknown required acks value: %s`, sinkCfg.RequiredAcks)
+	}
+
+	// TODO(#126991): Remove this sarama dependency.
+	// NOTE: kgo lets you give multiple compression options in preference order, which is cool but the config json doesnt support that. Should we?
+	switch sarama.CompressionCodec(sinkCfg.Compression) {
+	case sarama.CompressionNone:
+		opts = append(opts, kgo.ProducerBatchCompression(kgo.NoCompression()))
+	case sarama.CompressionGZIP:
+		opts = append(opts, kgo.ProducerBatchCompression(kgo.GzipCompression()))
+	case sarama.CompressionSnappy:
+		opts = append(opts, kgo.ProducerBatchCompression(kgo.SnappyCompression()))
+	case sarama.CompressionLZ4:
+		opts = append(opts, kgo.ProducerBatchCompression(kgo.Lz4Compression()))
+	case sarama.CompressionZSTD:
+		opts = append(opts, kgo.ProducerBatchCompression(kgo.ZstdCompression()))
+	default:
+		return nil, errors.Errorf(`unknown compression codec: %v`, sinkCfg.Compression)
+	}
+
+	if version := sinkCfg.Version; version != "" {
+		if !strings.HasPrefix(version, `v`) {
+			version = `v` + version
+		}
+		v := kversion.FromString(version)
+		if v == nil {
+			return nil, errors.Errorf(`unknown kafka version: %s`, version)
+		}
+		// NOTE: This version of kgo doesn't support specifying max versions
+		// >3.6.0 (released Oct 10 2023). This option is only really needed for
+		// interacting with <v0.10.0 clusters anyway.
+		opts = append(opts, kgo.MaxVersions(v))
+	}
+
+	return opts, nil
+}
+
+type kgoLogAdapter struct {
+	ctx context.Context
+}
+
+func (k kgoLogAdapter) Level() kgo.LogLevel {
+	if log.V(2) {
+		return kgo.LogLevelDebug
+	}
+	return kgo.LogLevelInfo
+}
+
+func (k kgoLogAdapter) Log(level kgo.LogLevel, msg string, keyvals ...any) {
+	format := `kafka(kgo): %s %s`
+	for i := 0; i < len(keyvals); i += 2 {
+		format += ` %s=%v`
+	}
+	log.InfofDepth(k.ctx, 1, format, append([]any{redact.SafeString(level.String()), redact.SafeString(msg)}, keyvals...)...) //nolint:fmtsafe
+}
+
+var _ kgo.Logger = kgoLogAdapter{}
+
+// newKgoChangefeedPartitioner returns a new kgo.Partitioner that wraps the
+// recommended sarama compat approach to pass thru record partitions when key is
+// nil, like the v1 implementation.
+func newKgoChangefeedPartitioner() kgo.Partitioner {
+	hasher := fnv.New32a()
+	return &kgoChangefeedPartitioner{
+		inner: kgo.StickyKeyPartitioner(kgo.SaramaCompatHasher(func(bs []byte) uint32 {
+			hasher.Reset()
+			_, _ = hasher.Write(bs)
+			return hasher.Sum32()
```

Comment:
Howdy, I'm looking more closely into twmb/franz-go#764 -- this code here is a race. It needs to be:

```go
hasher := fnv.New32a()
hasher.Write(bs)
return hasher.Sum32()
```

I just checked with `-gcflags '-m -m'` and this all inlines, so there is no new alloc for every partition call.

Reply:
Thanks! This should be fixed in https://github.com/cockroachdb/cockroach/pull/127585